dotnet/src/Extensions/PromptTemplates.Liquid/LiquidPromptTemplate.cs

```diff
@@ -116,4 +126,56 @@ public async Task<string> RenderAsync(Kernel kernel, KernelArguments? arguments
 
         return renderedResult;
     }
+
+    /// <summary>
+    /// Simple visitor for <see cref="ScriptPage"/> looking for variables that are only
+    /// ever read and appear to represent very simple strings.
+    /// </summary>
+    private sealed class FindVariablesVisitor : ScriptVisitor
+    {
+        private readonly HashSet<string> _variables = new(StringComparer.OrdinalIgnoreCase);
+        private bool _firstPass = true;
+
+        public static HashSet<string> FindInputs(Template template)
+        {
+            var visitor = new FindVariablesVisitor();
+            template.Page.Accept(visitor);
+            visitor._firstPass = false;
+            template.Page.Accept(visitor);
+            return visitor._variables;
```

Comment:
The first pass is adding things in, the second pass is taking things out. The first pass isn't adding things it knows aren't relevant, but it might add things that might not be relevant. For example, if you had:
```
{a} {b.a} {c.d}
```
the first pass is going to add `a` based on the `{a}`, but then it'll see the `{b.a}` and remove it; while processing the initial `{a}`, it didn't know it was going to hit the subsequent `{b.a}`.

This is just a simple heuristic and is far from ideal, but given the underlying library and what it provides, it's very non-trivial to make something more robust. I was actually thinking of pairing it back even further and only surfacing anything if everything is found to just be a non-complex top-level variable.

Reply:
Simplified.